Perform S3 directory deletion with batch requests

findinpath opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Description

<!-- Elaborate beyond the title of the PR as necessary to help the reviewers and maintainers.-->


Speed up the deletion of an S3 "directory" ( a path prefix which corresponds to multiple S3 objects) by using  batch deletion  requests.


<!-- Answer the following questions to help reviewers and maintainers
understand this PR's scope at a glance:
-->

> Is this change a fix, improvement, new feature, refactoring, or other?

Fix

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Hive, Delta, Iceberg (Lakehouse connectors)

> How would you describe this change to a non-technical end user or system administrator?

Speed up the deletion of an S3 "directory" ( a path prefix which corresponds to multiple S3 objects) by using  batch deletion  requests.


## Related issues, pull requests, and links

<!-- List any issues fixed by this PR, and provide links to other related PRs, upstream release notes, and other useful resources. For example:
* Fixes #issuenumber
* Related documentation in #issuenumber
* [Some release notes](http://usefulinfo.example.com)
-->

Fixes #13017 

<!-- The following sections are filled in by the maintainer with input from the contributor:
Use :white_check_mark: or (x) to signal selection.
-->

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Delta, Hive, Iceberg
* Ensure the rename/delete effectiveness for S3 directories which appear shallowly like objects
```

pettyjamesm commented:
I'm not sure this logic totally makes sense, because an S3 object with Content-Type: "application/octet-stream", with 0 bytes and a key that _doesn't_ end in `/` (eg: `s3://bucket/key`) seems like it should be interpreted as an "empty file" and not a directory. I could see the argument for interpreting an S3 object `s3://bucket/key/` as a directory, or `s3://bucket/key` as a "directory" even if an no object with that key exists so long as an "child object" exists (eg: some object with key `s3://bucket/key/object`)- but when an S3 object actually exists without the trailing slash I think we have to interpret that as a "file" and not directory unless it's appropriately marked with the right content type.

Now with that said, it's important to remember that S3 is not a file system, it's an object store so it doesn't _really_ have directories. So there might be a better option for how to handle `delete(Path path, boolean recursive == true)`. In that case, instead of doing this which recursively generates S3 listing, S3 getObjectMetadata, and S3 deleteObject calls:
```
for (FileStatus file : listStatus(path)) {
  delete(file.getPath(), true);
}
```

You could instead do something like (simplified to ignore details):
```
Iterator<S3ObjectSummary> listings = S3Objects.withPrefix(s3, bucketName, keyFromPath(path) + "/").iterator();
Iterator<String> keys = Iterators.transform(listings, S3ObjectSummary::getKey);
Iterator<List<String>> keyBatches = Iterators.partition(keys, 1000);
while (keyBatches.hasNext()) {
  String[] keysInBatch = keyBatches.next().toArray(String[]::new);
  // TODO: handle MultiObjectDeleteException in case some deletes fail
  s3.deleteObjects(new DeleteObjectsRequest(bucketName).withKeys(keysInBatch));
}
```

Which will be much more efficient, faster, and completely remove all S3 objects with actual keys starting with the prefix `"s3://<bucket>/<path>/"`.

findinpath commented:
@pettyjamesm  I have addressed your comment.

As you mentioned offline there is definitely room for improvement in `io.trino.plugin.hive.s3.TrinoS3FileSystem#rename` . I would suggest addressing the refactoring of this method in a separate PR.

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -590,24 +597,89 @@ public boolean rename(Path src, Path dst)
-        for (FileStatus file : listStatus(path)) {
-            delete(file.getPath(), true);
+        String bucketName = getBucketName(uri);
+        Iterator<S3ObjectSummary> listings = S3Objects.withPrefix(s3, bucketName, keyFromPath(path) + PATH_SEPARATOR).iterator();
```
My fault here for using `S3Objects.withPrefix` in an attempt to be concise in my suggested code, we probably want to add a new `ListMode` enum for `KEYS_WITH_PREFIX` and switch this to use `listPath(path, OptionalInt.empty(), KEYS_WITH_PREFIX)` so that we reuse the existing implementation and get metrics like the number of S3 listing calls. I recommend the name `KEYS_WITH_PREFIX` because `RECURSIVE_ALL` would suggest that all "directories" be present in the enumeration, but they won't be- all _keys_ will be returned, but no "directories" that are themselves just a prefix to some child key.

For example, say we have keys:
- `s3://bucket/path/key-1`
- `s3://bucket/path/subpath/key-2`

`listPath("s3://bucket/", OptionalInt.empty(), RECURSIVE_ALL)` would be expected to return:
- `s3://bucket/path`
- `s3://bucket/key-1`
- `s3://bucket/subpath`
- `s3://bucket/subpath/key-2`

Whereas `listPath("s3://bucket/", OptionalInt.empty(), KEYS_WITH_PREFIX)` will only return the actual keys above.

findinpath replied:
Thank you for pointing out `listPath` code reuse.

`RECURSIVE_ALL` does not exist. 

However we already have `RECURSIVE_FILES_ONLY` which very likely fits this specific use case.

pettyjamesm replied:
Correct, `RECURSIVE_ALL` does not exist but would be the missing "quadrant" in the space of `(SHALLOW | RECURSIVE) * (ALL | FILES_ONLY)`. The implied semantics of such a mode would be to include intermediate "directories" that have no corresponding object/key, which isn't really what we care about for deletes or renames. 

So the proposed new mode should be `RECURSIVE` but _not_ filter out "directories" which have actual S3 object keys (either marked with the `application/x-directory` content-type or ending with the suffix `"_$folder$"`), which `RECURSIVE_FILES_ONLY` does [here](https://github.com/trinodb/trino/blob/f6cad5e9450df12ba9e10071522587b09abde7b7/plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java#L694,L697). Instead, I propose the name of that new listing mode emphasize that it lists _all keys with a given prefix_ which will include "directories", but not "directories" that have no corresponding S3 object.

findinpath replied:
@pettyjamesm  I follow the logic of the above statement.

However, on testing I noticed that that `RECURSIVE_FILES_ONLY` does  fit the requirements of this method.

See the tests:
- `testDeleteNonRecursivelyPrefixContainingMultipleObjectsShouldFail`
- `testDeleteRecursivelyPathContainingPrefixCorrespondingAlsoToAnObjectKey`

findinpath replied:
After discussing on Slack with @pettyjamesm I understand now that we can have:
- objects
- directory objects (with content type `application/x-directory`)
- virtual directories (which exist only because they hold an object or a directory object)

I followed the hints from @pettyjamesm  and refactored the PR to be able to deal with directory objects.

pettyjamesm reviewed (commented):
Added review comments, I think we probably want to reorganize the logic to check `recursive` explicitly instead of handling that by falling through, since that's now a much more significant factor in the implementation logic. Recursive deletes for "implied directories" (ie: key does not exist but is a prefix for keys that do) and for "actual directories" (key exists, but has content-type `DIRECTORY_MEDIA_TYPE`) can proceed, but "empty objects" must be considered files and should not allow recursive deletes.

dnanuti commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -590,24 +597,89 @@ public boolean rename(Path src, Path dst)
+        Iterator<String> objectKeys = Iterators.transform(listings, S3ObjectSummary::getKey);
+        Iterator<List<String>> objectKeysBatches = Iterators.partition(objectKeys, BATCH_DELETE_OBJECTS_REQEST_MAX_PAGE_SIZE);
+
+        List<String> objectKeysFailedToDelete = new ArrayList<>();
```
Nit: LinkedList instead of ArrayList

L.E: Should we consider using a Set instead?

dnanuti commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -228,6 +234,7 @@
     private static final Set<String> GLACIER_STORAGE_CLASSES = ImmutableSet.of(Glacier.toString(), DeepArchive.toString());
     private static final MediaType DIRECTORY_MEDIA_TYPE = MediaType.create("application", "x-directory");
     private static final String S3_DEFAULT_ROLE_SESSION_NAME = "trino-session";
+    private static final int BATCH_DELETE_OBJECTS_REQEST_MAX_PAGE_SIZE = 1000;
```
Non-blocking: Should this be configurable?

findinpath replied:
> when would a user want to set it?

I think there is no need in being able to customize this setting.

As can be seen from aws-sdk-java `DeleteObjectsRequest` javadoc, there can be specified up to `1000` keys:

https://github.com/aws/aws-sdk-java/blob/8f6c93e49024049d2e5d882a3e2cd0e99e55ee65/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/model/DeleteObjectsRequest.java#L24-L32

dnanuti commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -590,24 +597,89 @@ public boolean rename(Path src, Path dst)
+                    if (requesterPaysEnabled) {
+                        // TODO use deleteObjectRequest.setRequesterPays() when https://github.com/aws/aws-sdk-java/issues/1219 is fixed
+                        // currently the method exists, but is ineffective (doesn't set the required HTTP header)
+                        deleteObjectsRequest.putCustomRequestHeader(Headers.REQUESTER_PAYS_HEADER, Constants.REQUESTER_PAYS);
```
Nit: static imports, the naming is self-explanatory (same above)

findinpath commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -590,24 +597,89 @@ public boolean rename(Path src, Path dst)
-        for (FileStatus file : listStatus(path)) {
-            delete(file.getPath(), true);
+        String bucketName = getBucketName(uri);
+        Iterator<S3ObjectSummary> listings = S3Objects.withPrefix(s3, bucketName, keyFromPath(path) + PATH_SEPARATOR).iterator();
```
Good catch. This should be handled via `io.trino.plugin.hive.s3.TrinoS3FileSystem#listPath` after the refactoring of the PR.

findinpath commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -590,24 +597,89 @@ public boolean rename(Path src, Path dst)
+            if (!path.getName().isEmpty()) {
+                MediaType metadataContentType = MediaType.parse(metadata.getContentType());
+                if (!metadataContentType.is(DIRECTORY_MEDIA_TYPE)) {
+                    if (!listPath(path, OptionalInt.of(1), ListingMode.SHALLOW_ALL).hasNext()) {
```
> this check is essentially a NOOP
Do note that `listPath` uses `path` + `/` as base of the listing and not `path`.
I would argue that this check is justified.
I'll need however to add test cases which test against AWS S3 to prove my assumptions.

pettyjamesm replied:
Ah, fair point- I missed that.

findinpath commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -590,24 +597,89 @@ public boolean rename(Path src, Path dst)
+            }
+
+            // check if this path is a directory
+            Iterator<LocatedFileStatus> iterator = listPath(path, OptionalInt.of(1), ListingMode.SHALLOW_ALL);
```
Do note that listPath uses `path` + `/` as prefix of the listing and not `path`.

findinpath commented:
>  I think we probably want to reorganize the logic to check recursive explicitly

I agree. The current state of the code for the `delete` method is rather hard to follow. I'll do a subsequent refactoring.

@pettyjamesm I've added a set of tests against AWS S3 object storage in order to be able to test the assumptions from the code.

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -590,24 +595,88 @@ public boolean rename(Path src, Path dst)
-        for (FileStatus file : listStatus(path)) {
-            delete(file.getPath(), true);
+        String bucketName = getBucketName(uri);
+        Iterator<LocatedFileStatus> listings = listPath(path, OptionalInt.empty(), ListingMode.RECURSIVE_FILES_ONLY);
```
`ListingMode.RECURSIVE_FILES_ONLY` will filter out "directories" that have corresponding S3 keys/objects, which we do want to delete when performing a recursive deletion. See [my other comment](https://github.com/trinodb/trino/pull/13974#discussion_r964867348) about this.

electrum commented:
We’re planning to replace `TrinoS3FileSystem` with a native `TrinoFileSystem` implementation, which has an API designed for object stores. Iceberg and Delta Lake already use it. Maybe instead of improving cod that we’ll throw away, we should build the new implementation now?

findinpath commented:
@electrum I understand the direction of replacing `TrinoS3FileSystem` with a native `TrinoFileSystem` implementation.
However, my change is rather punctual (related to the `delete` method) and starting the endeavour of making a native `TrinoFileSystem` implementation would take much more time. 
Even if the bugfix is short to medium lived, I still see a benefit in landing this new approach.

dnanuti commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestTrinoS3FileSystemAwsS3.java:
```diff
@@ -0,0 +1,314 @@
+                        paths.add(file.getPath().toString());
+                    }
+                }
+                assertThat(paths).containsOnly("s3://%s/%s/%s".formatted(bucketName, prefix, filename));
```
Nit: extract in a method the S3 path building `"s3://%s/%s/%s".formatted(bucketName, prefix, filename)`

dnanuti commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestTrinoS3FileSystemAwsS3.java:
```diff
@@ -0,0 +1,314 @@
+        return config;
+    }
+
+    private static String randomSuffix()
```
Would it make sense to use a UUID instead of this custom implementation?

dnanuti commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestTrinoS3FileSystemAwsS3.java:
```diff
@@ -0,0 +1,314 @@
+                fs.createNewFile(new Path("s3://%s/%s".formatted(bucketName, prefix), filename));
+                List<String> paths = new ArrayList<>();
+                RemoteIterator<LocatedFileStatus> files = fs.listFiles(new Path("s3://%s/%s".formatted(bucketName, prefix)), false);
+                while (files.hasNext()) {
```
It would remove some repetitive code if we'd extract this while loop in a method that takes as parameter `RemoteIterator<LocatedFileStatus>` and returns `List<String>`. I'd also use LinkedList as implementation instead of ArrayList.

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -590,25 +595,118 @@ public boolean rename(Path src, Path dst)
+        if (metadata == null) {
+            // the bucket root does not exist
+            if (path.getName().isEmpty()) {
+                return false;
```
When performing a recursive delete at the bucket root, shouldn't we still delete the objects within the bucket?

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -590,25 +595,118 @@ public boolean rename(Path src, Path dst)
+            }
+
+            // check if this path is a directory
+            Iterator<ListObjectsV2Result> listingsIterator = listObjects(path, OptionalInt.of(1), false);
```
It seems like ideally we would avoid redundant listings here vs inside of `deleteObjects`

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -590,25 +595,118 @@ public boolean rename(Path src, Path dst)
+                    }
+
+                    // Delete the object which has a key corresponding to the prefix to be deleted
+                    deleteObject(keyFromPath(path));
```
Shouldn't we call `deleteObject(keyFromPath(path))` when the content type is `application/x-directory` or when the object name ends with `_$folder$` ?

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -590,25 +595,118 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
```
Ok, so here's the high level expected control flow that I _think_ correctly captures the contract implementation and is more straightforward to reason about (although I think more cleanup is possible):
```
ObjectMetadata metadata = getS3ObjectMetadata(path);
if (recursive) {
  int deletedCount = 0;
  if (metadata == null || path.getName().endsWith("_$folder$") || MediaType.parse(metadata.getContentType()).is(DIRECTORY_MEDIA_TYPE)) {
    // directory (including "virtual" when metadata == null)
    deletedCount += deletePrefix(path + PATH_SEPARATOR); // only delete keys with: "<prefix>/" not "<prefix>_sibbling"
  }
  if (metadata != null && deleteObject(keyFromPath(path)) {
    deletedCount++; // "directory" object successfully deleted
  }
  return deletedCount > 0; // must delete at least one object to count as successful
}
else {
  if (metadata == null) {
    if (path.getName().isEmpty()) {
      return false; // bucket root does not exist
    }
    if (listPath(path, OptionalInt.of(1), SHALLOW_ALL).hasNext()) {
      return false; // non-empty directory
    }
    return deleteObject(keyFromPath(path) + PATH_SEPARATOR); // delete empty directory entry if present
  }
  else if (path.getName().endsWith("_$folder$") || MediaType.parse(metadata.getContentType()).is(DIRECTORY_MEDIA_TYPE)) {
    if (listPath(path, OptionalInt.of(1), SHALLOW_ALL).hasNext()) {
      return false; // non-empty directory
    }
  }
  return deleteObject(keyFromPath(path)); // file or empty directory deletion
}
```

findinpath replied:
Thank you @pettyjamesm for coming up with this algorithm.

It makes indeed things a bit more clear when doing deletes.

I took the time to implement the algorithm and also add corresponding tests for AWS and also MinIO to make sure the functionality is provided.

findinpath commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -590,27 +595,158 @@ public boolean rename(Path src, Path dst)
+                    return false; // bucket root does not exist
+                }
+                if (key.endsWith(DIRECTORY_SUFFIX)) {
+                    Path strippedPath = new Path(stripDirectorySuffix(path.toString()));
```
I'm rather reluctant to do the stripping of the directory suffix, but it seems that S3 doesn't list anything under the prefix, when the prefix contains  the `_$folder$` placeholder.

cc @pettyjamesm  any ideas how to handle this use case more elegantly?

pettyjamesm replied:
Oof, you're right. Ok, so... let's say that we have a delete for `s3://bucket/root_$folder$`. It stands to reason that we don't need to do anything special for recursive deletes underneath the root (eg: `s3://bucket/root/subpath_$folder$`) but at the top level there's a question about how to proceed. I think it stands to reason that for the top-level delete case, we want to strip `_$folder$` both when checking for nested "files or directories"  (ie: fail to delete if non-empty and non-recursive) as well as when doing a recursive delete (ie: delete `s3://bucket/root_$folder$` as well as object prefixed by `s3://bucket/root/`).

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -596,27 +601,158 @@ public boolean rename(Path src, Path dst)
+                deleted = deletePrefix(path);
+            }
+            if (key.endsWith(DIRECTORY_SUFFIX)) {
+                Path strippedPath = new Path(stripDirectorySuffix(path.toString()));
```
Shouldn't this be `new Path(stripDirectorySuffix(path.toString()) + PATH_SEPARATOR);`? Seems like the intention for a case like: `s3://bucket/directory_$folder$` would be to recursively delete `s3://bucket/directory/` (note that the trailing slash is significant, we don't want to delete objects like `s3://bucket/directory_sibling.txt`)

findinpath replied:
See that in `deletePrefix(Path)` method is called `listObjects(prefix, OptionalInt.empty(), true);` to which is added `PATH_SEPARATOR`.

```
        String key = keyFromPath(path);
        if (!key.isEmpty()) {
            key += PATH_SEPARATOR;
        }
```

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -596,27 +601,158 @@ public boolean rename(Path src, Path dst)
+            if (s3DirectoryMetadata != null) {
+                if (MediaType.parse(s3DirectoryMetadata.getContentType()).is(DIRECTORY_MEDIA_TYPE) ||
+                        key.endsWith(DIRECTORY_SUFFIX)) {
+                    String directoryKey = (key.endsWith(DIRECTORY_SUFFIX) ? stripDirectorySuffix(key) : key) + PATH_SEPARATOR;
```
When keys end with `_$folder$` exist as directories, don't we want to delete that object specifically and not the "stripped" object equivalent? It seems like above on line 609, we would assign `s3DirectoryMetadata` to `s3://bucket/directory_$folder$/` (not stripped, but with trailing slash added).

findinpath replied:
> When keys end with _$folder$ exist as directories, don't we want to delete that object specifically and not the "stripped" object equivalent.

I tried to do this initially :) but S3 seems to be effective on doing the deletion only when deleting the stripped object key.

findinpath replied:
I have now two test cases covering this specific question:

- `testDeleteRecursivelyPrefixWithDirectorySuffixContainingMultipleObjects`
- `testDeleteRecursivelyPrefixCorrespondingToDirectoryObjectWithDirectorySuffixContainingMultipleObjects`

pettyjamesm replied:
> I tried to do this initially :) but S3 seems to be effective on doing the deletion only when deleting the stripped object key.

I took a look at the tests in question, seems like this is an artifact of the way the test is written and not really the S3 behavior we're trying to test.

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -596,27 +601,158 @@ public boolean rename(Path src, Path dst)
+        ObjectMetadata s3ObjectMetadata = getS3ObjectMetadata(path, bucketName, key);
+        ObjectMetadata s3DirectoryMetadata = null;
+        if (!key.isEmpty()) {
+            s3DirectoryMetadata = getS3ObjectMetadata(path, bucketName, key + PATH_SEPARATOR);
```
This logic seems off for two reasons:
- non-recursive delete call shouldn't need this extra `getObjectMetadata`
- recursive delete to `_$folder$` suffixed entries should have `stripDirectorySuffix` called on the, no?

findinpath replied:
> non-recursive delete call shouldn't need this extra getObjectMetadata

This seems is a particularity of AWS S3 (not present in MinIO).
`io.trino.plugin.hive.s3.TestTrinoS3FileSystemAwsS3#testDeleteNonRecursivelyPrefixCorrespondingToNamesakeObjectAndEmptyDirectory` showcases that for deleting a key corresponding to an object and a namesake  directory we need two calls.

pettyjamesm replied:
Yeah, this would be one of the somewhat incoherent case where a path could be simultaneously a "file" and "directory" based on how we're trying to map the semantics. For example:
- Objects `s3://bucket/object` and `s3://bucket/object/` both exist.  Is "object" a file? A folder? The filesystem logic allows both to be true depending on the operation in question. `FileSystem.getLocatedFileStatus` and `FileSystem.listLocatedStatus` would both succeed with different results (unless "object" is marked `application/x-directory` in which case they would both call it a directory). 
- Objects `s3://bucket/prefix.txt` and `s3://bucket/prefix.txt/suffix.txt` both exist. Is `prefix.txt` a file? Folder? Same issue, could be either depending on the operation and whether `prefix.txt` is marked with `application/x-directory`.

I'm not sure how much of the potential craziness we really want to handle here if no S3 file systems in the wild actually create objects with these patterns.

Here's perhaps a helpful assumping: the caller has some intention about what is being deleted. Either they expect to delete a "file" or expect to delete a "directory", and will almost certainly pass `recursive = true` for the directory case and `false` for the file case. I don't expect it to be common for a directory deletion to pass "recursive: false" (although we do have to handle that scenario when the object in question is unambiguously a directory and uphold the "must be non-empty" contract). On that assumption, I think we can avoid this extra metadata lookup when `recursive == false` and move it inside the branch handling recursive deletes, and ignore the schizophrenic `s3://bucket/key` and `s3://bucket/key/` co-existing scenario for simple deletions.

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -596,27 +601,158 @@ public boolean rename(Path src, Path dst)
+
+        if (recursive) {
+            boolean deleted = false;
+            if (s3ObjectMetadata == null ||
```
The condition here seems wrong, what if this is called with `s3://bucket/directory/file.txt` ? I would expect `s3DirectoryMetadata` (resulting from a call to `getObjectMetadata("s3://bucket/directory/file.txt/")` which returns null) to be null and therefore fall into a `deletePrefix` instead of throwing an exception (it's not a "directory" in this scenario).

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -596,27 +601,152 @@ public boolean rename(Path src, Path dst)
+                if (listPath(strippedPath, OptionalInt.of(1), ListingMode.SHALLOW_ALL).hasNext()) {
+                    throw new IOException("Directory " + path + " is not empty");
+                }
+                return deleteObject(stripDirectorySuffix(key) + PATH_SEPARATOR);
```
If we end up here and `s3ObjectMetadata != null`, then shouldn't we be issuing a `deleteObject(key)` instead? We know that `s3://bucket/directory_$folder$` exists and is the "marker" that indicates the path `s3://bucket/directory` should be interpreted as a folder- but the S3 object key that needs deletion does actually have `_$folder$` suffixed.

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -596,27 +601,152 @@ public boolean rename(Path src, Path dst)
             }
+            if (s3DirectoryMetadata != null) {
+                String directoryKey = key + PATH_SEPARATOR;
+                deleted = deleteObject(directoryKey) || deleted;
```
Might make the logic slightly less brittle to use `s3DirectoryMetadata.getKey()` so that we aren't reconstructing the key via concatenation in the original lookup and subsequent delete.

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -596,27 +601,152 @@ public boolean rename(Path src, Path dst)
+                if (listPath(path, OptionalInt.of(1), ListingMode.SHALLOW_ALL).hasNext()) {
+                    throw new IOException("Directory " + path + " is not empty");
+                }
+                deleted = deleteObject(key + PATH_SEPARATOR);
```
Same thing here about preferring `s3DirectoryMetadata.getKey()`

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -596,27 +601,152 @@ public boolean rename(Path src, Path dst)
+                }
+                deleted = deleteObject(key + PATH_SEPARATOR);
+            }
+            if (s3ObjectMetadata != null) {
```
What if the object mime-type is `application/x-directory`? We would need the same "is empty" check before deletion, right?

findinpath commented:
After reviewing together with @pettyjamesm the implementation we've came with a simplified approach for the `delete(Path, boolean)` operation.

The previous approach which was dependent on analyzing metadata, has been replaced with another approach based primarily on listing the objects  and in case of dealing with recursive deletes, to perform the deletes in batches.

findinpath commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -596,27 +602,118 @@ public boolean rename(Path src, Path dst)
+            else {
+                deleteObject(key);
+            }
+            deleteObject(key + DIRECTORY_SUFFIX);
```
This may seem out of place because it makes sense only for directories.
However, listing with the prefix "directory/" under MinIO (oposed to AWS S3) doesn't list the "directory/" object key even if it exists. I've decided to perform two `deleteObject` operations instead of trying to figure out whether I have or not a directory by retrieving the object metadata call.

pettyjamesm replied:
Wouldn't the MinIO issue also affect the recursive case above?

findinpath replied:
Interestingly enough, if the directory object ( containing `/`) is created and has objects within it, it becomes in MinIO an virtual object and is not listed anymore. When deleting the objects it contains, the "base" directory dissapears as well.

findepi replied:
It looks like the dockerized MinIO != S3 (from behavioral perspective)

"s3-compatible storage" is a promise, a hope.

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -596,27 +602,118 @@ public boolean rename(Path src, Path dst)
+                return true;
+            }
+            else {
+                return deleteObject(key);
```
In this path, if `key` exists as an object with content type `application/x-directory` then we would need to call `deleteObject(key + DIRECTORY_SUFFIX)` as well. (will require making a `getObjectMetadata(path, bucket, key)` call).

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -596,27 +602,118 @@ public boolean rename(Path src, Path dst)
+                deleteObject(directoryKey);
+            }
+            else {
+                deleteObject(key);
```
The previous behavior would have preferred to consider this path to be a "file" and only delete this key if a non-directory object (ie: not marked with content-type: `application/x-directory`) with name `key` existed. When "files" are deleted, we also didn't previously issue `deleteObject(key + DIRECTORY_SUFFIX)` here.

findinpath replied:
I slightly modified the code to always delete the key with `DIRECTORY_SUFFIX` as well.

The option of using `listObjects` for gathering the keys to be deleted makes the code easy to follow. However as we observe, the fact that `MinIO` treats the "directory" object keys and the listing differently compared to AWS S3 makes this option rather wobbly :( .

findinpath replied:
I did this change to cover the MinIO empty directory usecase.
The directory will not get listed in this situation for deletion, and needs to be deleted in the same fashion as an "object". Note that using key corresponding to the provided path (without a trailing `/`) works in MinIO to delete an "empty directory" object.

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -596,27 +602,118 @@ public boolean rename(Path src, Path dst)
+    @VisibleForTesting
+    static Iterator<String> keysFromListing(ListObjectsV2Result listing)
+    {
+        List<String> prefixes = listing.getCommonPrefixes();
```
Returning `prefixes` as "keys" in `keysFromListing` seems wrong, because they aren't actually S3 objects. We are expecting to have done a "no path separator" listing which should never include prefix entries in the response. Maybe we should refactor away this code or fail loudly if prefixes are present?

pettyjamesm reviewed (commented):
Looks much cleaner with the simplifications from before, added some further comments.

findinpath commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -596,27 +602,118 @@ public boolean rename(Path src, Path dst)
+        if (recursive) {
+            boolean deleted = deletePrefix(path);
+            if (deleted) {
+                deleteObject(key + DIRECTORY_SUFFIX);
```
If the key exists as on object (no trailing `/`), `deleted` should be `false` and the object will be deleted in the `else` branch.

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -596,27 +603,108 @@ public boolean rename(Path src, Path dst)
-        for (FileStatus file : listStatus(path)) {
-            delete(file.getPath(), true);
+        // Clean up eventual objects which the deletion has failed while doing the batched deletion
+        while (!objectKeysFailedToDelete.isEmpty()) {
```
Why would we retry (twice) here?

findinpath replied:
I think I've overengineered  a bit here. Dropping the retry mechanism altogether.

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -596,27 +603,108 @@ public boolean rename(Path src, Path dst)
+            try {
+                DeleteObjectsRequest deleteObjectsRequest = new DeleteObjectsRequest(bucketName).withKeys(objectKeysBatch);
+                if (requesterPaysEnabled) {
+                    // TODO use deleteObjectRequest.setRequesterPays() when https://github.com/aws/aws-sdk-java/issues/1219 is fixed
+                    // currently the method exists, but is ineffective (doesn't set the required HTTP header)
+                    deleteObjectsRequest.putCustomRequestHeader(REQUESTER_PAYS_HEADER, REQUESTER_PAYS);
```
The linked issue suggests this works correctly for DeleteObjectsRequest (plural)

findepi commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestTrinoS3FileSystemAwsS3.java:
```diff
@@ -0,0 +1,54 @@
+    protected Configuration defaultS3Configuration()
+    {
+        Configuration config = newEmptyConfiguration();
+        config.set("fs.s3a.endpoint", s3Endpoint);
```
s3a -> s3 ?

findepi commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestTrinoS3FileSystemMinio.java:
```diff
@@ -0,0 +1,181 @@
+
+        minioClient =
+                MinioClient.builder()
+                        .endpoint("http://localhost:" + minio.getMinioApiEndpoint().getPort())
```
why ignore minio.getMinioApiEndpoint().getHost()?

btw see [Make it easier to use MinIO with Hadoop](https://github.com/trinodb/trino/pull/14489/commits/f4fa22f0aa2b5ecc3ec4a28aa7310aa34a0d43a6) in https://github.com/trinodb/trino/pull/14489

findepi commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestTrinoS3FileSystemMinio.java:
```diff
@@ -0,0 +1,181 @@
+    {
+        String testBucketName = "trino-" + randomUUID();
+        minioClient.makeBucket(MakeBucketArgs.builder().bucket(testBucketName).build());
+        String testBucketPath = "s3://%s".formatted(testBucketName);
```
we don't use format when value is simply appended

findinpath commented on plugin/trino-hive/pom.xml:
```diff
@@ -521,5 +522,22 @@
         </profile>
+
+        <profile>
+            <id>test-hive-s3</id>
```
We don't actually need a new profile. I was thinking that `test-hive-glue` is not appropriate for doing s3 tests.
Should i rename the profile to a more generic name? `test-hive-aws` ?

findepi replied:
`aws-tests`

findinpath commented:
Rebasing on top of `master` to make use of the changes from https://github.com/trinodb/trino/pull/14489

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -598,25 +604,76 @@ public boolean rename(Path src, Path dst)
-                return deleteObject(keyFromPath(path));
+        String key = keyFromPath(path);
+        if (recursive) {
+            boolean deleted = deletePrefix(path);
```
I think the logic around `deletePrefix(path)` is sketchy here, because `deletePrefix` will return `false` if any objects failed to delete. I guess before we also would have returned `true` if any of the individual deletes failed, so nothing ultimately has changed- but below we decide whether to call `deleteObject(key)` based on whether _everything_ was deleted as expected. That's maybe justifiable, but then we should return `false` from this method if the intention was to leave the directory in tact. So here are the options that seem reasonable to me:
- Remove the `if (!deleted)` check guarding `deleteObject(key)`, call it unconditionally regardless of whether there were sub-objects deleted. Either continue always returning `true` from the method, or return whether any of the deletes deleted at least one value.
- Have some return value signal for failures having occurred in `deletePrefix` and avoid calling `deleteObject(key)` or `deleteObject(key + DIRECTORY_SUFFIX)` when failures occurred, return false from the method. Alternatively, proceed with the attempt to `deleteObject(key)` and `deleteObject(key + DIRECTORY_SUFFIX)` regardless of failures and return `true` if at least one of those three deletes succeeded.
-

findinpath replied:
I've refactored the logic to take indeed note of this change. Sorry for overseeing it while removing the retries.

findepi replied:
ultimately, `io.trino.filesystem.TrinoFileSystem#deleteDirectory` doesn't return a `boolean` -- it should throw if anything left behind

for `org.apache.hadoop.fs.FileSystem#delete(org.apache.hadoop.fs.Path, boolean)` we should return true if and only if everything deleted, i.e. return false  if anything left behind

pettyjamesm commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -598,25 +604,76 @@ public boolean rename(Path src, Path dst)
+                s3.deleteObjects(deleteObjectsRequest);
+            }
+            catch (MultiObjectDeleteException e) {
+                e.getErrors().stream().map(MultiObjectDeleteException.DeleteError::getKey).forEach(objectKeysFailedToDelete::add);
```
With the logic now no longer retrying failed deletes we can discard the `Set<String>` in favor of a `boolean someDeletesFailed` or `int objectsNotDeleted` instead.

findinpath commented:
Updated the PR from `master` to use the latest MinIO image for testing which offers increased S3 API compatibility .

findepi commented:
the build seems red

krvikash commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestTrinoS3FileSystemMinio.java:
```diff
@@ -0,0 +1,143 @@
+
+            AmazonS3 s3 = fs.getS3Client();
+            String filename1 = "file1.txt";
+            fs.createNewFile(new Path("s3://%s/%s".formatted(testBucketName, filename1)));
```
nit: I would prefer to use a string literal when the variable is not going to use more than once.

findepi commented:
@findinpath please rebase

findinpath commented:
Rebased to deal with conflicts from `master`.

findepi commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/AbstractTestTrinoS3FileSystemObjectStorage.java:
```diff
@@ -0,0 +1,546 @@
+    }
+
+    @Test
+    public void testDeleteRecursivelyPrefixCorrespondingToDirectoryObjectContainingMultipleObjectsSpread()
```
as in other comment -- try to make test method names a bit shorter, while still meaningful

findinpath commented:
> `BATCH_DELETE_OBJECTS_REQEST_MAX_PAGE_SIZE`
> when would a user want to set it?

I think there is no need in being able to customize this setting.

As can be seen from aws-sdk-java `DeleteObjectsRequest` javadoc, there can be specified up to `1000` keys:

https://github.com/aws/aws-sdk-java/blob/8f6c93e49024049d2e5d882a3e2cd0e99e55ee65/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/model/DeleteObjectsRequest.java#L24-L32

findinpath commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -606,25 +610,86 @@ public boolean rename(Path src, Path dst)
+                if (requesterPaysEnabled) {
+                    deleteObjectsRequest.setRequesterPays(true);
+                }
+                s3.deleteObjects(deleteObjectsRequest);
```
As per `DeleteObjectsResult` javadoc this object is being returned in case that all keys submitted for deletion where successfully deleted. 

https://github.com/aws/aws-sdk-java/blob/8f6c93e49024049d2e5d882a3e2cd0e99e55ee65/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/model/DeleteObjectsResult.java#L23-L30

findinpath commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -606,25 +610,86 @@ public boolean rename(Path src, Path dst)
+                deleteObject(directoryKey);
+            }
+            else {
+                deleteObject(key);
```
However, in order to keep things relatively simple this S3 filesystem implementation considers that this situation is not supposed to appear in real world in context of data lake tables.

This is why, if the path provided doesn't correspond to a "directory" prefix, we'll attempt to delete the object corresponding to the specified key.

findepi replied:
I think i didn't get a straight answer to

> why is deleteObject(key) attempted only if key + '/' object does not exist?

let's document the reasoning as a code comment

findinpath commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/AbstractTestTrinoS3FileSystemObjectStorage.java:
```diff
@@ -0,0 +1,546 @@
+                String directoryPath = "s3://%s/%s".formatted(getBucketName(), directoryKey);
+                createDirectory(fs.getS3Client(), getBucketName(), directoryKey);
+
+                String filename1 = "file1.txt";
+                String filename2 = "file2.txt";
+                String filename3 = "file3.txt";
+                String filename4 = "file4.txt";
+                fs.createNewFile(new Path(directoryPath, filename1));
+                fs.createNewFile(new Path(directoryPath, filename2));
+                fs.createNewFile(new Path(directoryPath, filename3));
+                fs.createNewFile(new Path(directoryPath, filename4));
```
One does suffice.
I've simplified the test and also the subsequent test.

findinpath commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/AbstractTestTrinoS3FileSystemObjectStorage.java:
```diff
@@ -0,0 +1,546 @@
+    private static final MediaType DIRECTORY_MEDIA_TYPE = MediaType.create("application", "x-directory");
+    private static final String PATH_SEPARATOR = "/";
+    private static final String DIRECTORY_SUFFIX = "_$folder$";
+
```
Added:

- testDeleteNonRecursivelyDeletesOnlyRequestedObject
- testDeleteNonRecursivelyDeletesOnlyRequestedDirectory
- testDeleteRecursivelyDeletesOnlyRequestedObject
- testDeleteRecursivelyDeletesOnlyRequestedDirectory

findinpath commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/AbstractTestTrinoS3FileSystemObjectStorage.java:
```diff
@@ -0,0 +1,546 @@
+    }
+
+    @Test
+    public void testDeleteRecursivelyPrefixContainingOneObject()
```
At the moment of this writing this test case is not anymore special.
In the first iterations of the PR however the implementation was different and  more test cases were needed to cover different flows in the deletion logic. 
I removed this test along a few others which have become irrelevant.

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -606,25 +610,80 @@ public boolean rename(Path src, Path dst)
+                deleteObject(childKey);
+            }
+            else {
+                if (isNotEmpty(key)) {
```
Does this mean the object referenced by the key is not empty (more than 0 bytes)?

findinpath replied:
By object is meant the java `String` object?
I think the `key` can be either empty `""` or non empty (indicating an object within the s3 bucket) because it is extracted from a `Path`.

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -169,6 +171,7 @@
 import static java.util.concurrent.Executors.newCachedThreadPool;
 import static java.util.concurrent.TimeUnit.SECONDS;
 import static java.util.stream.Collectors.toList;
+import static org.apache.commons.lang3.StringUtils.isNotEmpty;
```
avoid

findepi replied:
Prefer JDK and Guava. The two have generally consistent API rules, especially around null-friendliness.
Apache Commons has a different null-friendliness philosophy, and it should be avoided in the project.

findinpath replied:
I replaced the check with `String.isBlank`

findepi commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestTrinoS3FileSystemMinio.java:
```diff
@@ -0,0 +1,161 @@
+    public void testDeleteNonRecursivelyEmptyBucketRoot()
+            throws Exception
+    {
+        String testBucketName = "trino-delete-bucket-root-empty";
```
Due to filesystem caches this may need to be randomized

https://github.com/trinodb/trino/blob/af73f3b4834a7d52bd2d15bf34d61d4f871f8672/testing/trino-testing-containers/src/main/java/io/trino/testing/minio/MinioClient.java#L162-L167

findinpath replied:
I opted for creating a `randomBucketSuffix()` method inspired from 

https://github.com/trinodb/trino/blob/f60bc22016b10793518d7d87e13faf797d6df868/testing/trino-testing/src/main/java/io/trino/testing/sql/TestTable.java#L133-L137

Using `UUID.randomUUID()` doesn't work because MinIO limits the lenght of the buckets to be between `3` and `63` characters.

findepi commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestTrinoS3FileSystemMinio.java:
```diff
@@ -0,0 +1,175 @@
+        }
+    }
+
+    private static String randomBucketSuffix()
```
In other places we shamelessly abuse `randomTableSuffix`

https://github.com/trinodb/trino/blob/389403af94c28e3c6c17ebdca0c95bae2bd170e4/plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseTestHiveOnDataLake.java#L79

instead of adding (many) clones of this method, let's rename it -- https://github.com/trinodb/trino/pull/15008

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDropTableCompatibility.java:
```diff
@@ -61,7 +61,8 @@ public static Object[][] engineConfigurations()
                 {DELTA, DELTA, true},
-                {DELTA, DELTA, false},
+                {DELTA, TRINO, false},
+                {DELTA, DELTA, false}
```
it's resolved, but the comma is missing :)

findinpath replied:
I felt there is no more need to add `,` to support adding eventually in the future another combination, because all the engine configurations are already listed. Nevertheless, I'm adding the extra `,`  to be in sync with the existing code in Trino.

findepi commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/AbstractTestTrinoS3FileSystemObjectStorage.java:
```diff
@@ -0,0 +1,546 @@
+    }
+
+    @Test
+    public void testDeleteRecursivelyPrefixContainingMultipleObjectsSpread()
```
"nested"?

findepi commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestTrinoS3FileSystemMinio.java:
```diff
@@ -0,0 +1,175 @@
+    public static final String MINIO_SECRET_KEY = "secretkey";
+
+    private static final SecureRandom random = new SecureRandom();
+    // The suffix needs to be long enough to "prevent" collisions in practice. The length of 5 was proven not to be long enough
```
nvrm, just remove it and use https://github.com/trinodb/trino/pull/15008

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -606,25 +609,81 @@ public boolean rename(Path src, Path dst)
+                deleteObject(childKey);
+            }
+            else {
+                // Avoid deleting the bucket in case that the provided path points to the bucket root
```
I don't think we would delete a bucket, would we?

Maybe 
_Do not attempt to "delete the bucket", of the path is the bucket root, as this would fail_

findepi replied:
is it true for S3, or another MinIO quirk?

findinpath replied:
I just tried an adaptation of the test `io.trino.plugin.hive.s3.TestTrinoS3FileSystemMinio.testDeleteNonRecursivelyEmptyBucketRoot` on an AWS S3  while having the check for the empty key commented out and it turns out that also on AWS S3 the bucket is being removed.

> is it true for S3, or another MinIO quirk?

Yes, it is true for both `S3` and `MinIO`.

findepi commented:
Test PR with secrets: https://github.com/trinodb/trino/pull/15098
(@nineinchnick's https://github.com/trinodb/trino/pull/12817 will make these PRs easier)

Also, the first commit is separately being tested here: https://github.com/trinodb/trino/pull/15099

findepi commented:
@findinpath can you please investigate the CI outcome?

findinpath commented:
https://github.com/trinodb/trino/actions/runs/3495735811/jobs/5852837911 showcases that the content can't be deleted on `presto-ci-test` S3 bucket. 

```
Error:  io.trino.plugin.hive.s3.TestTrinoS3FileSystemAwsS3.testDeleteNonRecursivelyNonEmptyDeepPath  Time elapsed: 0.403 s  <<< FAILURE!
java.io.IOException: Failed to delete paths with the prefix path s3://***/test-illegal-delete-non-recursively-deep-path-non-empty-e69bb17a-8e62-49e2-b281-1a84037c5992
...
Caused by: com.amazonaws.services.s3.model.AmazonS3Exception: Access Denied (Service: Amazon S3; Status Code: 403; Error Code: AccessDenied; Request ID: EHCHYPMFXNKTEW9C; S3 Extended Request ID: OfGRQM1L2Kv6gX2pQLjTiLWxdmEJCUkHa5nkpa5qrtyXSaqZVlpb9X6R8TgIYP+h8Tm8gW3fFf4=; Proxy: null), S3 Extended Request ID: OfGRQM1L2Kv6gX2pQLjTiLWxdmEJCUkHa5nkpa5qrtyXSaqZVlpb9X6R8TgIYP+h8Tm8gW3fFf4=
```

findinpath commented:
Rebasing on top of `master` to make use of `TRINO_AWS_ACCESS_KEY_ID` & `TRINO_AWS_SECRET_ACCESS_KEY` secrets

findinpath commented:
Rebased on top of `master` in order to cope with the fact that the hive AWS tests are using now different secrets, reason why I needed to correspondingly change the S3 bucket used for testing to `trino-ci-test`.
The only file changed after the rebasing was `ci.yml`

ebyhr commented:
Updated the mirror PR #15098

findinpath commented:
CI hit https://github.com/trinodb/trino/issues/13199 for `ci / pt (default, suite-delta-lake-databricks113`.

alexjo2144 commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -606,25 +609,81 @@ public boolean rename(Path src, Path dst)
+            }
+            else {
+                // Avoid deleting the bucket in case that the provided path points to the bucket root
+                if (!key.isEmpty()) {
```
Can we move this check to an `key.isEmpty()` early return as the first thing in the not recursive block?

findinpath replied:
Moving the check up would cause the method to return `false` while trying to delete empty keys even when the bucket is not empty - which would not hold the contract of the operation.
See `io.trino.plugin.hive.s3.TestTrinoS3FileSystemMinio#testDeleteNonRecursivelyNonEmptyBucketRoot`.

alexjo2144 commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:
```diff
@@ -606,25 +609,81 @@ public boolean rename(Path src, Path dst)
+                        .withQuiet(true));
+            }
+            catch (AmazonS3Exception e) {
+                log.debug(e, "Failed to delete objects from the bucket %s under the prefix '%s'", bucketName, prefix);
```
`log.error`? or at least warn

findinpath replied:
https://github.com/trinodb/trino/pull/13974#discussion_r1015229654

findepi commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/BaseTestTrinoS3FileSystemObjectStorage.java:
```diff
@@ -403,11 +397,7 @@ public void testDeleteRecursivelyDirectoryWithObjectsAndDirectorySuffixPlacehold
-                assertThat(listPaths(fs.getS3Client(), getBucketName(), prefix, true))
-                        .containsOnly(
-                                directoryKey + "/",
-                                directoryKey + "/dir4/");
```
Does this correspond to https://github.com/trinodb/trino/issues/13017?
if so, can you place "https://github.com/trinodb/trino/issues/13017" in the TODO line?

findepi commented:
`ci / check-commits (pull_request) Failing after 37s`

findinpath commented:
Rebasing on top of `master` to address `check-commits` issue.

findepi commented:
/test-with-secrets sha=12265ea06d6ac9e006bb56973cb23cc54a8e9d2d

findepi commented:
(rebased to fix test-with-secrets job)

findepi commented:
/test-with-secrets sha=07e689f46a2a40fe6e618b02a0eae4a2ff492b76

findepi commented:
Run with secrets: https://github.com/trinodb/trino/actions/runs/3591234020

findepi commented:
@findinpath what were the changes in the last two force-pushes (https://github.com/trinodb/trino/pull/13974#event-7930164952, https://github.com/trinodb/trino/pull/13974#event-7932782205)?

findinpath commented:
I have changed the `setup()` and `tearDown()` methods from `TestTrinoS3FileSystemMinio` so that there are no resources allocated in the constructor.

findepi commented:
/test-with-secrets sha=ba2d0b67d0890174483733fe48012373c7478ead

findepi commented:
/test-with-secrets sha=62a44129acd1d4db0e8630062e3c01901bac0882

findinpath commented:
CI Hit #13199

findepi commented:
@nineinchnick do you understand the `Pull Request Labeler / pt with secrets (pull_request_target)` failure  (https://github.com/trinodb/trino/pull/13974/checks?check_run_id=9998437275)?
